pkg/oauthflow/device.go

```diff
@@ -0,0 +1,157 @@
+
+const (
+	/* #nosec */
+	SigstoreDeviceURL = "https://oauth2.sigstore.dev/auth/device/code"
+	/* #nosec */
+	SigstoreTokenURL = "https://oauth2.sigstore.dev/auth/device/token"
+)
+
+type deviceResp struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationUri         string `json:"verification_uri"`
+	VerificationUriComplete string `json:"verification_uri_complete"`
+	Interval                int    `json:"interval"`
+}
+
+type tokenResp struct {
+	IDToken string `json:"id_token"`
+	Error   string `json:"error"`
+}
+
+type DeviceFlowTokenGetter struct {
+	MessagePrinter func(string)
+	Sleeper        func(time.Duration)
+	Issuer         string
+	CodeURL        string
+	TokenURL       string
+}
+
+func NewDeviceFlowTokenGetter(issuer, codeURL, tokenURL string) *DeviceFlowTokenGetter {
+	return &DeviceFlowTokenGetter{
+		MessagePrinter: func(s string) { fmt.Println(s) },
+		Sleeper:        time.Sleep,
+		Issuer:         issuer,
+		CodeURL:        codeURL,
+		TokenURL:       tokenURL,
+	}
+}
+
+func (d *DeviceFlowTokenGetter) deviceFlow(clientID string) (string, error) {
+	data := url.Values{
+		"client_id": []string{clientID},
+		"scope":     []string{"openid email"},
+	}
+
+	/* #nosec */
+	resp, err := http.PostForm(d.CodeURL, data)
+	if err != nil {
+		return "", err
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	parsed := deviceResp{}
+	if err := json.Unmarshal(b, &parsed); err != nil {
+		return "", err
+	}
+	d.MessagePrinter(fmt.Sprintf("Open your browser to: %s", parsed.VerificationUriComplete))
+
+	for {
+		// Some providers use a secret here, we don't need for sigstore oauth one so leave it off.
+		data := url.Values{
+			"grant_type":  []string{"urn:ietf:params:oauth:grant-type:device_code"},
+			"device_code": []string{parsed.DeviceCode},
+			"scope":       []string{"openid", "email"},
+		}
+
+		/* #nosec */
+		resp, err := http.PostForm(d.TokenURL, data)
+		if err != nil {
+			return "", err
+		}
+		b, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return "", err
+		}
+		tr := tokenResp{}
+		if err := json.Unmarshal(b, &tr); err != nil {
+			return "", err
+		}
+
+		if tr.IDToken != "" {
+			d.MessagePrinter("Token received!")
+			return tr.IDToken, nil
+		}
+		switch tr.Error {
+		case "access_denied", "expired_token":
+			return "", fmt.Errorf("error obtaining token: %s", tr.Error)
+		case "authorization_pending":
+		case "slow_down":
+		default:
+			d.Sleeper(time.Duration(parsed.Interval) * time.Second)
+			continue
+		}
+	}
+}
+
+func (d *DeviceFlowTokenGetter) GetIDToken(p *oidc.Provider, cfg oauth2.Config) (*OIDCIDToken, error) {
+
+	idToken, err := d.deviceFlow(cfg.ClientID)
+	if err != nil {
+		return nil, err
+	}
+	verifier := p.Verifier(&oidc.Config{ClientID: cfg.ClientID})
+	parsedIDToken, err := verifier.Verify(context.Background(), idToken)
+	if err != nil {
+		return nil, err
+	}
+
+	var claims struct {
```

Comment:
can use `EmailFromIDToken` instead of duplicating the logic

Reply:
EmailFromIDToken actually disappeared somehow. Added it back